"Rule of 5" copy/assign/move declarations

cary-ilm opened the pull request:
- Added =delete constructor/operator=/move to all internal classes (declared inside a .cpp)
- Added =delete constructor/operator=/move to all external classes (in .h files) for which the operation is not trivial (i.e. class has dynamically-allocated fields), and is currently not needed.
- Added =default constructor/operator=/move to TypedAttribute<T> class, since that behavior is reasonable.
- added if(this!=&other) in operator=
- restructured IMF_STD_ATTRIBUTE

cary-ilm committed:
sonar fixes

Signed-off-by: Cary Phillips <cary@ilm.com>

cary-ilm committed:
sonar fixes

Signed-off-by: Cary Phillips <cary@ilm.com>

cary-ilm committed:
SonarCloud-inspired fixes

- Added =delete copy constructors and assignment operators
- Added =default destructors
- check for &other==this in assignment operator

Signed-off-by: Cary Phillips <cary@ilm.com>

cary-ilm committed:
More SonarCloud-inspired fixes:

- copy constructors and assignment operators declared = delete
- use of ## is macro definition

Signed-off-by: Cary Phillips <cary@ilm.com>

cary-ilm committed:
SonarCloud-inspired bug fixes:

- copy constructor and assignment operator declared = delete
- compare this==&other in assignment operator

Signed-off-by: Cary Phillips <cary@ilm.com>

cary-ilm committed:
SonarCloud-inspired fixes

- removed const from operator= =delete
- added move-assign/move-constructor = delete

Signed-off-by: Cary Phillips <cary@ilm.com>

xlietz commented:
You probably saw that this is generating an undefined move constructor in the Windows build when linking testMultiPartSharedAttributes.obj.  I suspect this is related to MSVC compiler generating a move constructor while gcc/clang are doing some optimization akin to copy elision. Possibly related to line 146 in testMultiPartSharedAttributes.cpp where header is an xvalue in the call to headers.push_back(header) but I haven't traced it to where the TypedAttr<int> move constructor would actually be invoked. Just a guess as I haven't had time to test it!

2019-10-25T22:00:12.5717991Z   testMultiPartSharedAttributes.obj : error LNK2019: unresolved external symbol "public: __thiscall Imf_2_4::TypedAttribute<int>::TypedAttribute<int>(class Imf_2_4::TypedAttribute<int> const &&)"
(??0?$TypedAttribute@H@Imf_2_4@@QAE@$$QBV01@@Z) referenced in function "private: class Imf_2_4::TypedAttribute<int> * __thiscall std::vector<class Imf_2_4::TypedAttribute<int>,class std::allocator<class Imf_2_4::TypedAttribute<int> > >::_Umove(class Imf_2_4::TypedAttribute<int> *,class Imf_2_4::TypedAttribute<int> *,class Imf_2_4::TypedAttribute<int> *)" (?_Umove@?$vector@V?$TypedAttribute@H@Imf_2_4@@V?$allocator@V?$TypedAttribute@H@Imf_2_4@@@std@@@std@@AAEPAV?$TypedAttribute@H@Imf_2_4@@PAV34@00@Z) [D:\a\1\s\_build\OpenEXR\IlmImfTest\IlmImfTest.vcxproj]
2019-10-25T22:00:12.5718387Z   D:\a\1\s\_build\bin\Release\IlmImfTest.exe : fatal error LNK1120: 1 unresolved externals [D:\a\1\s\_build\OpenEXR\

cary-ilm commented:
I guess in this case there's no reason to not provide the move constructor as well.  These changes are taking the conservative approach - if the function isn't declared, eliminated it, as long as there's no legitimate need for it, but this presumably points to a need for it. It's worth investigating a little more, but I'll just add the move constructor.

xlietz commented:
These copy/move deletions break ABI compatibility. And possibly API compatibility when both are deleted. Just to verify, we are ok with that? Any lib user who has been invoking these copy/move constructors intentionally or unintentionally, will not be able to compile, Perhaps not a bad thing, they can add the constructor if there is legitimate need?

lgritz commented:
They are all things that shouldn't be copied or moved, so I'm ok with it.

But the potential break break in ABI compatibility means it should not be backported to 2.4.x for sure (I'm ok with ABI breaks for minor releases such as 2.4 -> 2.5; IIRC, semver only talks about API, not ABI, so I think it's not a break of semver rules for 2.5).

The potential break in API compatibility means it's probably a 3.0 feature if you want to follow strict sementic versioning.

cary-ilm committed:
=default copy/move/assign for TypedAttribute

TypedAttribute assumes that the T class supports copy/move/assign
operations, so it should, too.

Also, the move/moveassign operators for OStream/IStream incorrectly
declared their arguments as const.

Signed-off-by: Cary Phillips <cary@ilm.com>

cary-ilm commented:
Changed the name of the PR and edited the description, since the majority of the fixes involve copy/move/assign operations.

cary-ilm committed:
remove const from arg declaration in move-constructors/move-assignments

Signed-off-by: Cary Phillips <cary@ilm.com>

cary-ilm committed:
mvoe TestType and TestTypedAttribute to OPENEXR_IMF_INTERNAL_NAMESPACE

Signed-off-by: Cary Phillips <cary@ilm.com>

cary-ilm committed:
typo from previous commit: operator= wasn't returning a value

Signed-off-by: Cary Phillips <cary@ilm.com>

meshula reviewed (approved):
checks out :)
